### PR TITLE
[DRAFT v1] Define the SPAKE2+ algorithms

### PR DESCRIPTION
**Pull request by athoelke**
[Edited] : This is the first draft. An revised and reorganised version of this PR is now available: #129 

---

This PR merges the 'best guess' result of the other PAKE and SPAKE2+ PRs (#114, #119, #120, #124), before defining the SPAKE2+ algorithms. 

The Algorithm definition begins in commit https://github.com/ARM-software/psa-api/commit/e784de8ef06984d518e00d139e2a22fe0effdcbb, in case you have reviewed the other pieces already.

For a draft of the whole spec: [Crypto_API-1.2_PAKE_Extension-bet.2-draft-spake2plus.pdf](https://github.com/ARM-software/psa-api/files/13370131/AES0058-PSA_Certified_Crypto_API-1.2_PAKE_Extension-bet.2-draft-spake2plus.pdf)

### DO NOT MERGE

There remain open issues in the preceding PRs. Once there is broad agreement on the open API design issues, I will merge the PR changes, to permit a second pass editorial review and rework of the result.

I am not convinced that the document structure is quite right. SPAKE2+ benefits from significant explanatory text for how the keys, algorithms, and operation work together - but this material is shared by three algorithm and two key API identifiers, and is currently somewhat scattered. This material cannot be delegated to other sources (as is the case for most other algorithms in the API), because the asymmetric key approach used here is novel, and we [reasonably] constrain some SPAKE2+ options in order to conform to the Matter specification.

### CODE REVIEW

**Comment by athoelke** [comment]
@silabs-Kusumit, @silabs-hannes, @oberon-sk, @yanesca: this might be the PR you have been waiting for?

**Comment by athoelke** [comment]
Reviewing the flow:

SPAKE2+ authentication flow graphic from the PR

It seems slightly odd that the key is provided during the `psa_pake_setup()`, and the role is set afterward, even though the role (for SPAKE2+) dictates what kind of key can be used. For example, if the operation is setup with a SPAKE2+ public key, then a subsequent call to `psa_pake_set_role(..., PSA_PAKE_ROLE_CLIENT)` would have to fail with `PSA_ERROR_INVALID_ARGUMENT`.

It seems it might feel more logical for the role to be declared as part of the setup (see #115 where the key input was moved to the setup function) so the combination of cipher-suite, key and role can be validated together.

On the other hand, it is possible to call `psa_pake_set_role()` with other roles (e.g. `PSA_PAKE_ROLE_FIRST`), that are invalid for the cipher-suite. So maybe this is not as odd as it first seemed to me?

**Comment by silabs-Kusumit** [comment]
I think that `psa_pake_set_role()` should be a separate function from `psa_pake_setup()`. The `set_role()` function is similar to `set_user()`, `set_peer()` functions and it doesn't feel right to add `set_role` to the `setup` function for only one PAKE algorithm.

**Comment by silabs-Kusumit** [comment]
> I am not convinced that the document structure is quite right. SPAKE2+ benefits from significant explanatory text for how the keys, algorithms, and operation work together - but this material is shared by three algorithm and two key API identifiers, and is currently somewhat scattered. This material cannot be delegated to other sources (as is the case for most other algorithms in the API), because the asymmetric key approach used here is novel, and we [reasonably] constrain some SPAKE2+ options in order to conform to the Matter specification.

Maybe adding a document in the appendix folder for the details on SPAKE2+ would be helpful

**Comment by athoelke** [comment]
> Maybe adding a document in the appendix folder for the details on SPAKE2+ would be helpful

I've thought about this a bit more, and think there are approximately three main ways to slice this:

  1. Split the API definition into three parts: SPAKE2+ section, JPAKE section, common section.

     The algorithm-specific sections would provide an introduction to the protocol, how to use the API for the protocol (the flows), the key types, formats & derivations; and the algorithm identifiers. The common section would provide the definition of the shared parts of the API (primitives, cipher-suites, roles, and the operation itself).

     Given that the protocol descriptions only make proper sense with an understanding of the common API, it might be better to define the common section first, and then introduce the protocols?

  2. Maintain something like the first draft, where the how-to for a protocol is described as part of the various elements that make it up.

  3. Move the how-to material to an appendix.

The last option would mostly work in this Extension specification. However, when integrated into the main specification, I would be concerned that they would become detached from the PAKE API definition, and lose value as a result.

On the whole I have been trending toward moving the summaries for each category of algorithm into the section that defines that API. For example, the change in v1.1 where most of that material from the Functionality overview was moved into the intro to each cryptographic operation section of the API.

I'm going to draft something like (1) above for the next revision, to see how that looks...

File: doc/ext-pake/api/pake.rst [hunk 1]
```diff
@@ -61,6 +165,146 @@ Changes and additions to the Programming API
...
+        The elliptic curve family id, if ``type`` is a supported SPAKE2+ key. Unspecified if ``type`` is not a supported SPAKE2+ key.
+
+.. _spake2p-key-derivation:
+
+Derivation of SPAKE2+ keys
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The SPAKE2+ key types can be output from a key derivation using :code:`psa_key_derivation_output_key()`. The SPAKE2+ protocol recommends that a key-stretching kdf, such as PBKDF2, is used to hash the SPAKE2+ password. See RFC 9383 for details.
+
+For example, after setting up the PBKDF2 operation, the following process will derive the SPAKE2+ key pair for use with the P-256 Elliptic curve group :issue:`(This example may be more than necessary in the specification?)`:
```
[the remark this answers is not in the document]
Comment (athoelke) [reply doc/ext-pake/api/pake.rst:265]: This is what I have done in Draft-2, as part of the reorganisation.

File: doc/ext-pake/api/pake.rst [hunk 2]
```diff
@@ -61,6 +165,146 @@ Changes and additions to the Programming API
...
+
+For the |API|:
+
+*   The derivation of SPAKE2+ keys extracts ceil(log2(p)/8) + 8 bytes from the PBKDF for each of w0s and w1s, where p is the prime factor of the order of the elliptic curve group.
+*   The calculation of w0, w1, and L then proceeds as described in the RFC.
+*   A SPAKE2+ key-pair is the pair (w0, w1).
+*   A SPAKE2+ public key is the pair (w0, L).
+
+.. todo::
+    Would a table of required w0s/w1s lengths for each of the supported SPAKE2+ elliptic curve groups be useful here?
```
[the remark this answers is not in the document]
Comment (athoelke) [reply doc/ext-pake/api/pake.rst:307]: I have done this in the updated #119, and so in Draft-2 (#129).

**Comment by athoelke** [comment]
This PR has been superseded by the alternative document layout in the [v2] PR #129.

**Pull request closed without merging**